worker/upload.go

```diff
@@ -0,0 +1,836 @@
+package worker
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"math"
+	"sort"
+	"sync"
+	"time"
+
+	"go.opentelemetry.io/otel/attribute"
+	"go.opentelemetry.io/otel/trace"
+	rhpv2 "go.sia.tech/core/rhp/v2"
+	"go.sia.tech/core/types"
+	"go.sia.tech/renterd/api"
+	"go.sia.tech/renterd/object"
+	"go.sia.tech/renterd/tracing"
+	"lukechampine.com/frand"
+)
+
+// TODO: add priority queues
+
+const defaultAvgUpdateInterval = 10 * time.Second
+
+var errNoQueue = errors.New("no queue")
+
+type (
+	uploadID [8]byte
+
+	uploader struct {
+		w *worker
+
+		statsOverdrive *average
+		statsStopChan  chan struct{}
+
+		poolMu sync.Mutex
+		pool   []*uploadQueue
+		bh     uint64
+	}
+
+	uploadQueue struct {
+		u  *uploader
+		hp hostProvider
+		rl revisionLocker
+
+		fcid       types.FileContractID
+		hk         types.PublicKey
+		siamuxAddr string
```

Comment:
The pool gets updated on every upload. 99% of the time nothing will happen but if the contract set changed the upload is removed from the pool and a new one is added (potentially). I think a host re-announcing is currently not covered but that's tricky, do we need to cover that you think? It will get updated on the next scan and until then it'll just drop down to the bottom of the pool and be used as few times as possible.